source/crud/crud.rst

```diff
@@ -1953,54 +1953,37 @@ in the collection prior to ``$merge`` being executed).
+
 Historically, only primaries could execute an aggregation pipeline with ``$out``
-or ``$merge``. As of ``featureCompatibilityVersion`` 4.4, secondaries can now
-execute such pipelines; however, since drivers do not track
-``featureCompatibilityVersion``, executing an aggregation pipeline with ``$out``
-or ``$merge`` on a secondary will require a 5.0+ server (i.e. wire version >=
-13).
-
-Drivers MUST guarantee the following:
-
-- If an explicit (i.e. per-operation) read preference is specified for an
-  aggregation with a write stage, drivers MUST attempt to use it. If that would
-  result in a pre-5.0, secondary server being selected, drivers MUST instead
-  select a server using a primary read preference.
-- If no explicit read preference is specified but a default read preference is
-  available to inherit (e.g. from the Collection), drivers MUST attempt to use
-  it. If that would result in a pre-5.0, secondary server being selected,
-  drivers MUST instead select a server using a primary read preference.
-
-Drivers SHOULD use a custom server selector to consider server/wire version when
-matching a read preference and, if a pre-5.0 secondary would be selected, fall
-back to selecting a primary. With this approach, only a single server selection
-attempt is needed.
-
-If it is not possible to augment the server selector and a pre-5.0 secondary is
-selected, drivers MAY invoke server selection a second time to obtain a primary.
-However, if the first attempt results in a server selection timeout, drivers
-MUST NOT make a second attempt and MUST propagate the original timeout error. If
-the topology type is Single, Sharded, or LoadBalanced, drivers MUST NOT make a
-second attempt (there is no benefit in doing so).
-
-Regardless of whether drivers employ a custom server selector or two-attempt
-approach, they MUST discern the *effective* read preference for the operation,
-which SHALL be used for specifying the
+or ``$merge`` and drivers never considered a read preference for the operation.
+As of ``featureCompatibilityVersion`` 4.4, secondaries can now execute pipelines
+with ``$out`` or ``$merge``. Since drivers do not track
+``featureCompatibilityVersion``, the decision to consider a read preference for
+such a pipeline will depend on the wire version(s) of the server(s) to which the
+driver is connected.
+
+If there are one or more available servers and one or more of those servers is
+pre-5.0 (i.e. wire version < 13), drivers MUST NOT use the available read
+preference and MUST instead select a server using a primary read preference.
+Otherwise, if there are either no available servers or all available servers are
+5.0+ (i.e. wire version >= 13), drivers MUST use the available read preference.
+
+Drivers SHOULD augment their
+`server selection algorithm <..../server-selection/server-selection.rst#server-selection-algorithm>`_
+such that this logic can be enforced within a single server selection attempt.
```

Comment:
I'm no longer going into detail about the two-attempt approach, since it may only apply to PHP and the the original text conflicted with the new rule to consider _all_ servers to which the driver is connected (to determine any evidence of a connection to a pre-5.0 server).

In the previous text, the two-attempt approach instructed the driver to fall back to a primary only if the available RP selected a pre-5.0 secondary. With the new rule, I think the two attempt approach becomes significantly more complicated:

 * Attempt to select a server using the available RP. If this is the first operation executed by the driver, this first server selection attempt will initialize SDAM and allow us to check all available servers in the next step.
 * Regardless of whether server selection succeeds or fails, check the wire version of all available servers
 * If there is no evidence of a pre-5.0 server, either use the server we selected or propagate the server selection failure.
 * If there is any evidence of a pre-5.0 server, note that `{mode: "primary"}` is the effective read preference for this operation (i.e. we must "fall back"). If we already happened to select a primary in the first attempt, we can use it without making a second selection attempt. Otherwise, we must invoke server selection again to select a primary.

One notable difference here is that we can no longer abort if the first attempt raises a server selection failure. This change in behavior should also address the BC concern I raised in [DRIVERS-1969](https://jira.mongodb.org/browse/DRIVERS-1969?focusedCommentId=4161573&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4161573):

> Row 6 (P4, S-none): P4 will be selected for most RPs; however, RP(secondary) will result in a server selection failure because a pre-5.0 secondary is never selected to warrant fall back behavior. This is potentially a behavioral BC break

Reply:
Did you figure out a way to still use the two-attempt approach in PHP without changes to libmongoc API? Or will libmongoc API also be altered as a result of this spec change?

Reply:
I'd rather create a convoluted approach in PHPLIB than introduce API changes in both libmongoc _and_ PHPC just so PHPLIB can utilize them. My current plan is to attempt server selection with the read preference and then regardless of whether a server is selected or we get an error, I'll check all available servers for pre-5.0 and conditionally select a primary if needed.